registry/storage/driver/gcs/gcs.go

```diff
@@ -751,57 +751,66 @@ func (d *driver) Delete(context context.Context, path string) error {
 		}
 		return nil
 	}
-	err = storageDeleteObject(gcsContext, d.bucket, d.pathToKey(path))
-	if err != nil {
-		if status, ok := err.(*googleapi.Error); ok {
-			if status.Code == http.StatusNotFound {
-				return storagedriver.PathNotFoundError{Path: path}
-			}
-		}
+	err = storageDeleteObject(ctx, d.bucket, d.pathToKey(path), d.gcs)
+	if err == storage.ErrObjectNotExist {
+		return storagedriver.PathNotFoundError{Path: path}
 	}
 	return err
 }
 
-func storageDeleteObject(context context.Context, bucket string, name string) error {
-	return retry(func() error {
-		return storage.DeleteObject(context, bucket, name)
-	})
+func storageDeleteObject(ctx context.Context, bucket string, name string, gcs *storage.Client) error {
+	return gcs.Bucket(bucket).Object(name).Delete(ctx)
 }
 
-func storageStatObject(context context.Context, bucket string, name string) (*storage.Object, error) {
-	var obj *storage.Object
+func (d *driver) storageStatObject(ctx context.Context, name string) (*storage.ObjectAttrs, error) {
+	bkt := d.gcs.Bucket(d.bucket)
+	var obj *storage.ObjectAttrs
 	err := retry(func() error {
 		var err error
-		obj, err = storage.StatObject(context, bucket, name)
+		obj, err = bkt.Object(d.pathToKey(name)).Attrs(ctx)
 		return err
 	})
 	return obj, err
 }
 
-func storageListObjects(context context.Context, bucket string, q *storage.Query) (*storage.Objects, error) {
-	var objs *storage.Objects
-	err := retry(func() error {
-		var err error
-		objs, err = storage.ListObjects(context, bucket, q)
-		return err
-	})
-	return objs, err
+func storageListObjects(ctx context.Context, bucket string, q *storage.Query, gcs *storage.Client) ([]*storage.ObjectAttrs, error) {
+	bkt := gcs.Bucket(bucket)
+	var objs []*storage.ObjectAttrs
+	it := bkt.Objects(ctx, q)
```

Comment:
If we _do_ need to keep a single slice (but see my other comment), would it be possible to initialise `objs` with an initial capacity based on that (to reduce allocations)?

```go
objs := make([]*storage.ObjectAttrs, 0, `<some expected size>`)
```

Not sure what information we have available about the expected number of results though; I see the iterator has a PageInfo, but that only has the number of items remaining before a new API call has to be made; 
https://pkg.go.dev/google.golang.org/api/iterator#PageInfo.Remaining